pydicom/charset.py

```diff
@@ -51,18 +51,72 @@
     'GBK': 'GBK',  # from DICOM correction CP1234
 }
 
+# Map Python encodings to escape sequences as defined in PS3.3 in tables
+# C.12-3 (single-byte) and C.12-4 (multi-byte character sets).
+escape_codes = {
+    'ascii': b'\x1b(B',  # used to switch to ASCII G0 code element
+    'iso8859': b'\x1b-A',
+    'shift_jis': b'\x1b)I',
+    'iso2022_jp': b'\x1b$B',
+    'iso8859_2': b'\x1b-B',
+    'iso8859_3': b'\x1b-C',
+    'iso8859_4': b'\x1b-D',
+    'iso_ir_126': b'\x1b-F',
+    'iso_ir_127': b'\x1b-G',
+    'iso_ir_138': b'\x1b-H',
+    'iso_ir_144': b'\x1b-L',
+    'iso_ir_148': b'\x1b-M',
+    'iso_ir_166': b'\x1b-T',
+    'euc_kr': b'\x1b$)C',
+    'iso-2022-jp': b'\x1b$(D',
+    'iso_ir_58': b'\x1b$)A 58',
+}
+
 default_encoding = "iso8859"
 
+ESCAPE = b'\x1b'
 
-def clean_escseq(element, encodings):
-    """Remove escape sequences that Python does not remove from
-       Korean encoding ISO 2022 IR 149 due to the G1 code element.
+
+def decode_string(value, encodings):
+    """Convert a raw byte string into a unicode string using the given
+    list of encodings.
     """
-    if 'euc_kr' in encodings:
-        return element.replace("\x1b\x24\x29\x43", "").replace(
-            "\x1b\x28\x42", "")
+    if ESCAPE not in value:
+        return value.decode(encodings[0])
+
+    # multi-byte character sets except Korean are handled by Python encodings
+    use_python_handling = value.startswith((escape_codes['iso2022_jp'],
+                                            escape_codes['iso-2022-jp'],
+                                            escape_codes['iso_ir_58']))
+
+    if use_python_handling:
+        parts = [value]
     else:
-        return element
+        # Each part of the value that starts with an escape sequence is
+        # decoded separately using the corresponding encoding.
+        # See PS3.5, 6.1.2.4 and 6.1.2.5 for the use of code extensions.
+        parts = [ESCAPE + part for part in value.split(ESCAPE) if part]
+        # the first part may not start with an escape sequence
+        if not value.startswith(ESCAPE):
+            parts[0] = parts[0][1:]
+    result = u''
+    for part in parts:
+        if part.startswith(ESCAPE):
+            for enc in list(encodings) + ['ascii']:
+                if enc in escape_codes and part.startswith(escape_codes[enc]):
+                    if use_python_handling:
+                        val = part.decode(enc)
+                    else:
+                        val = part[len(escape_codes[enc]):].decode(enc)
+                    break
+            else:
+                # unknown escape code - use first encoding (probably incorrect)
+                val = part.decode(encodings[0])
```

Comment:
Should we have a warning message if that's the case?

Reply:
I will handle the warnings later in a separate PR (there are quite a bit more).